app/controllers/admin/product_import_controller.rb

```diff
@@ -1,62 +1,95 @@
-  end
+    def validate_data
+      return unless process_data('validate')
+      render json: @importer.import_results, response: 200
+    end
 
-  private
+    def save_data
+      return unless process_data('save')
+      render json: @importer.save_results, response: 200
+    end
+
+    def reset_absent_products
+      @importer = ProductImport::ProductImporter.new(File.new(params[:filepath]), spree_current_user, import_into: params[:import_into], enterprises_to_reset: params[:enterprises_to_reset], updated_ids: params[:updated_ids], settings: params[:settings])
+
+      if params.key?(:enterprises_to_reset) && params.key?(:updated_ids)
+        @importer.reset_absent(params[:updated_ids])
+      end
 
-  def validate_upload_presence
-    unless params[:file] || (params[:filepath] && File.exist?(params[:filepath]))
-      redirect_to '/admin/product_import', notice: I18n.t(:product_import_file_not_found_notice)
-      return
+      render json: @importer.products_reset_count
     end
-  end
 
-  def check_file_errors(importer)
-    if importer.errors.present?
-      redirect_to '/admin/product_import', notice: @importer.errors.full_messages.to_sentence
-      return
+    private
+
+    def validate_upload_presence
+      unless params[:file] || (params[:filepath] && File.exist?(params[:filepath]))
+        redirect_to '/admin/product_import', notice: I18n.t(:product_import_file_not_found_notice)
+      end
     end
-  end
 
-  def check_spreadsheet_has_data(importer)
-    unless importer.item_count
-      redirect_to '/admin/product_import', notice: I18n.t(:product_import_no_data_in_spreadsheet_notice)
-      return
+    def process_data(method)
+      @importer = ProductImport::ProductImporter.new(File.new(params[:filepath]), spree_current_user, start: params[:start], end: params[:end], settings: params[:settings])
+
+      begin
+        @importer.send("#{method}_entries")
+      rescue StandardError => e
+        render json: e.message, response: 500
+        return false
+      end
+
+      true
     end
-  end
 
-  def save_uploaded_file(upload)
-    filename = 'import' + Time.now.strftime('%d-%m-%Y-%H-%M-%S')
-    extension = '.' + upload.original_filename.split('.').last
-    directory = 'tmp/product_import'
-    Dir.mkdir(directory) unless File.exists?(directory)
-    File.open(Rails.root.join(directory, filename+extension), 'wb') do |f|
-      f.write(upload.read)
-      f.path
+    def check_file_errors(importer)
+      if importer.errors.present?
+        redirect_to '/admin/product_import', notice: @importer.errors.full_messages.to_sentence
+      end
     end
-  end
 
-  # Define custom model class for Cancan permissions
-  def model_class
-    ProductImporter
+    def check_spreadsheet_has_data(importer)
+      unless importer.item_count
+        redirect_to '/admin/product_import', notice: I18n.t(:product_import_no_data_in_spreadsheet_notice)
+      end
+    end
+
+    def save_uploaded_file(upload)
```

Comment:
Perhaps not now, but I feel this controller has way too many responsibilities and it would benefit from extracting some into new objects/s.

Reply:
What do you think @Matt-Yorkley?